### PR TITLE
Translate otlp metrics to prom metatada

### LINKED ISSUE #4910: OTLP Metrics don't send metadata
#### Describe the bug
When ingesting OTLP metrics, we drop the metadata (HELP and TYPE) without propagating it.

#### To Reproduce
Push OTLP Metrics. 

Currently the translation function from upstream OTel Collector doesn't return metadata so we just ignore it: https://github.com/grafana/mimir/blob/4d4a9a9d6bc3a9799e254fe55c7094cf3326d35d/pkg/util/push/otel.go#LL128C1-L128C1

#### Expected behavior

We also append the metadata.

cc @kovrus

### PR DESCRIPTION
**Pull request by jmichalek132**
#### What this PR does

Translate OTLP to Metadata and forward it. 
This makes metrics description and type available when ingest OTLP protocol.

#### Which issue(s) this PR fixes or relates to

Fixes #4910

#### Checklist

- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by gouthamve** [comment]
https://github.com/grafana/mimir/issues/4910#issuecomment-1668452485

1. Enhancement is correct.
2. Yes, but lets do it in a different PR?
3. Going to defer to the maintainers here. But I _think_ so.
4. You can extend it.

**Comment by jmichalek132** [comment]
* Updated tests, please let me know if this is good enough
* Added TODO for exposing addMetricSuffixes to the configuration used [here](https://github.com/grafana/mimir/pull/5693/files#diff-eb47b1b5619b45185602515d8d5274a06451058d43aaa9c85c9fc1a7222635c3R176)
* Waiting for the decision on  feature flag/configuration option to enable this

**Comment by jmichalek132** [comment]
Hi, @gouthamve  @charleskorn could I please bother you for a review?

File: CHANGELOG.md
Comment (charleskorn) [inline CHANGELOG.md:69]:
```suggestion
* [ENHANCEMENT] Distributor: store metadata when ingesting metrics via OTLP. This makes metrics description and type available when ingest OTLP protocol. #5693
```

File: integration/e2emimir/client.go
Comment (charleskorn) [inline integration/e2emimir/client.go:295]: Why not just do `url.Values{}` here?

File: integration/e2emimir/client.go
Comment (charleskorn) [inline integration/e2emimir/client.go:303]: Don't the query parameters (`body`) need to be included in the URL, not the request body?

File: integration/e2emimir/client.go
Comment (charleskorn) [inline integration/e2emimir/client.go:312]: I don't think this comment is necessary.

File: integration/otlp_ingestion_test.go
Comment (charleskorn) [inline integration/otlp_ingestion_test.go:109]: Might be worth adding a check that the metadata is stored correctly for this histogram series too.

File: pkg/util/push/otel.go
Comment (charleskorn) [inline pkg/util/push/otel.go:178]: I'd suggest extracting `scopeMetrics.Metrics().At(k)` out to a variable to reduce the duplication here.

File: pkg/util/push/otel.go
Comment (charleskorn) [inline pkg/util/push/otel.go:176]: Is this TODO still relevant?

**Review by charleskorn — COMMENTED** [review]
Thanks for putting this together @jmichalek132. 

As you suggested in [your comment on the issue](https://github.com/grafana/mimir/issues/4910#issuecomment-1668452485), I think it would be good to add a configuration flag to allow enabling / disabling this. For now, let's mark it as experimental and default to disabled until we've had a chance to test this at scale.

I'd also like to get a review from someone more familiar with OTLP before we merge, as I'm not an OTLP expert.

File: pkg/util/push/otel.go
Comment (jmichalek132) [reply pkg/util/push/otel.go]: Done.

File: integration/e2emimir/client.go
Comment (jmichalek132) [reply integration/e2emimir/client.go]: Removed

**Comment by jmichalek132** [comment]
Is there a way to run the linter that will also apply the formatting?

File: integration/e2emimir/client.go
Comment (jmichalek132) [reply integration/e2emimir/client.go]: Removed

File: integration/otlp_ingestion_test.go
Comment (jmichalek132) [reply integration/otlp_ingestion_test.go:109]: Added.

File: pkg/util/push/otel.go
Comment (jmichalek132) [reply pkg/util/push/otel.go]: Yes as per @gouthamve request, exposing this boolean into configuration should be done in separate PR
https://github.com/grafana/mimir/pull/5693#issuecomment-1669216300
Answering my question
https://github.com/grafana/mimir/issues/4910#issuecomment-1668452485

File: pkg/util/push/otel.go
Comment (charleskorn) [inline pkg/util/push/otel.go:48]: The convention in Go is for parameter names to be camelCase.

```suggestion
	enableOtelMetadataStorage bool,
```

File: pkg/util/push/push_test.go
Comment (charleskorn) [inline pkg/util/push/push_test.go:1]: Could you please add a test for the case where the feature flag is disabled?

File: pkg/api/api.go
Comment (charleskorn) [inline pkg/api/api.go:53]: I don't think this comment is necessary, we have the description on the flag to help people understand what this is for.

File: pkg/api/api.go
Comment (charleskorn) [inline pkg/api/api.go:73]: [nit]

```suggestion
	f.BoolVar(&cfg.EnableOtelMetadataStorage, "distributor.enable-otlp-metadata-storage", false, "If true, store metadata when ingesting metrics via OTLP. This makes metric descriptions and types available for metrics ingested via OTLP.")
```

File: pkg/util/push/otel.go
Comment (charleskorn) [inline pkg/util/push/otel.go:167]: One small performance improvement here would be to pre-allocate a slice large enough to hold all of the expected metadata for this request.

**Comment by charleskorn** [comment]
> Is there a way to run the linter that will also apply the formatting?

`golangci-lint run --fix ./...` will automatically fix some issues.

File: pkg/util/push/otel.go
Comment (jmichalek132) [reply pkg/util/push/otel.go]: fixed

File: pkg/api/api.go
Comment (jmichalek132) [reply pkg/api/api.go]: removed

**Comment by colega** [comment]
The CHANGELOG has just been cut to prepare for the next Mimir release. Please rebase `main` and eventually move the CHANGELOG entry added / updated in this PR to the top of the CHANGELOG document. Thanks!

File: pkg/util/push/otel.go
Comment (jmichalek132) [reply pkg/util/push/otel.go]: updated

File: pkg/api/api.go
Comment (charleskorn) [inline pkg/api/api.go:53]: Could you please add this to the list of experimental features in [the docs](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md)?

File: pkg/util/push/otel.go
Comment (charleskorn) [inline pkg/util/push/otel.go:184]: Is this TODO something you're planning to address in this PR?

Comment (jmichalek132) [reply pkg/util/push/otel.go:184]: No, @gouthamve asked to leave the implementation of that for follow up PR here https://github.com/grafana/mimir/pull/5693#issuecomment-1669216300, should I remove the comment and create an issue for it?

File: pkg/api/api.go
Comment (jmichalek132) [reply pkg/api/api.go:53]: will do

File: pkg/util/push/otel.go
Comment (gouthamve) [reply pkg/util/push/otel.go:184]: I'd create an issue yup.

Comment (jmichalek132) [reply pkg/util/push/otel.go:184]: Created https://github.com/grafana/mimir/issues/5967. Should I remove the comment or leave it in?

File: pkg/api/api.go
Comment (jmichalek132) [reply pkg/api/api.go:53]: done, is it ok?

File: pkg/util/push/otel.go
Comment (gouthamve) [reply pkg/util/push/otel.go:184]: I'd put #5967 in the comment tbh. I'd leave it in.

File: pkg/util/push/push_test.go
Comment (jmichalek132) [reply pkg/util/push/push_test.go:1]: Added here https://github.com/grafana/mimir/pull/5693/files#diff-9cfabdded42ea41657229be85019e9543e391f591bc1aaee09fd36d8b83fc218R130 is this okay?

File: CHANGELOG.md
Comment (charleskorn) [inline CHANGELOG.md:11]:
```suggestion
* [ENHANCEMENT] Distributor: store metadata when ingesting metrics via OTLP. This makes metrics description and type available when ingesting metrics via OTLP. #5693
```

**Review by charleskorn — APPROVED** [review]
LGTM, thanks @jmichalek132!

File: CHANGELOG.md
Comment (charleskorn) [inline CHANGELOG.md]:
```suggestion
* [ENHANCEMENT] Distributor: add experimental support for storing metadata when ingesting metrics via OTLP. This makes metrics description and type available when ingest OTLP protocol. Enable with `-distributor.enable-otlp-metadata-storage=true`. #5693
```

File: CHANGELOG.md
Comment (charleskorn) [inline CHANGELOG.md:11]:
```suggestion
* [ENHANCEMENT] Distributor: add experimental support for storing metadata when ingesting metrics via OTLP. This makes metrics description and type available when ingesting metrics via OTLP. Enable with `-distributor.enable-otlp-metadata-storage=true`. #5693
```

**Comment by charleskorn** [comment]
Could you please fix the linting and test failures @jmichalek132?

**Comment by jmichalek132** [comment]
> Could you please fix the linting and test failures @jmichalek132?

All should be fixed now. Can you please approve the test run again?

**Review by charleskorn — APPROVED** [review]

**Comment by charleskorn** [comment]
@gouthamve would you mind taking another look at this when you get a chance?

File: pkg/util/push/otel.go
Comment (gouthamve) [inline pkg/util/push/otel.go:183]: I'm confused here. `metric.Type()` returns this: https://pkg.go.dev/go.opentelemetry.io/collector/pdata/pmetric#MetricType

While https://github.com/grafana/mimir/blob/main/pkg/mimirpb/mimir.pb.go#L53-L64 is a different thing. Specifically OTLP has "Sum" while Mimir has "COUNTER". How are we mapping them together?

Comment (jmichalek132) [reply pkg/util/push/otel.go:183]: Yeah I will look into how is this handled in https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/prometheusexporter

Comment (jmichalek132) [reply pkg/util/push/otel.go]: Added the function for translating OTEL metric types to Prometheus metric types.

File: pkg/util/push/otel.go
Comment (jmichalek132) [inline pkg/util/push/otel.go:182]: Should I add warning / debug log message here when the default case happens?

File: pkg/util/push/otel.go
Comment (gouthamve) [inline pkg/util/push/otel.go:182]:
```suggestion
	case pmetric.MetricTypeGauge:
		return mimirpb.GAUGE
	case pmetric.MetricTypeSum:
		metricType := mimirpb.GAUGE
		if otelMetric.Sum().IsMonotonic() {
			metricType = mimirpb.COUNTER
		}
		return metricType
	case pmetric.MetricTypeHistogram:
		return mimirpb.HISTOGRAM
	case pmetric.MetricTypeSummary:
		return mimirpb.SUMMARY
	case pmetric.MetricTypeExponentialHistogram:
		return mimirpb.HISTOGRAM
	}
	return mimirpb.UNKNOWN
```

File: pkg/util/push/push_test.go
Comment (gouthamve) [inline pkg/util/push/push_test.go]: Similar changes here.

Comment (jmichalek132) [reply pkg/util/push/push_test.go]: fixed

**Review by gouthamve — APPROVED** [review]
LGTM! @charleskorn This is ready to go imo!

**Pull request merged**
